apps/decoder_csv/lib/decoder/csv.ex

```diff
@@ -26,24 +26,48 @@ defmodule Decoder.Csv do
     def lines_or_bytes(_t), do: :line
 
     def decode(t, messages) do
-      if skip_line?(t) do
-        messages
-        |> Enum.drop(1)
-        |> Enum.map(fn message -> parse(message, t.headers) end)
-      else
-        Enum.map(messages, fn message -> parse(message, t.headers) end)
-      end
+      do_decode(t, messages, t.skip_first_line && !already_skipped?())
     end
 
-    def skip_line?(%{skip_first_line: false}), do: false
+    defp do_decode(t, messages, false = _should_skip?) do
+      Enum.map(messages, &parse(&1, t.headers))
+    end
+
+    defp do_decode(t, messages, true = _should_skip?) do
+      mark_skipped()
+
+      messages
+      |> Enum.drop(1)
+      |> Enum.map(fn message -> parse(message, t.headers) end)
+    end
```

Comment:
Curious what @bbalser thinks about `do_decode` just outputting the messages unmodified if `_should_skip? = false` and `messagees |> Enum.drop(1)` if `_should_skip? = true` and then having the output of `do_decode` just piped to `Enum.map(fn message -> parse(message, t.headers end)` since lines 33 and 41 are identical.

Reply:
I pushed a commit that does this. I dig it.